### PR TITLE
mgr/dashboard: provide hub Cluster HA for multi-cluster setup

### PR DESCRIPTION
**Pull request by aaSharma14**
When adding a cluster to the multi-cluster setup, set all the mgr IP's as cross_origin_url in the connected cluster to provide HA for the hub cluster in case of a mgr failover

Fixes: https://tracker.ceph.com/issues/65499

[Screencast from 2024-04-12 11-16-07.webm](https://github.com/ceph/ceph/assets/66050535/dc8dc11b-da8f-456a-8ffd-f3e62b3bebba)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 1]
```diff
@@ -101,6 +105,19 @@ def auth(self, url: str, cluster_alias: str, username: str,
...
 
+    def get_cors_endpoints_string(self, hub_url):
+        parsed_url = urlparse(hub_url)
+        hostname = parsed_url.hostname
+        cors_endpoints_list = []
+        orch = OrchClient.instance()
+        inventory_hosts = [host.to_json() for host in orch.hosts.list()]
+        for host in inventory_hosts:
+            host_addr = host['addr']
+            host_url = hub_url.replace(hostname, host_addr)
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:116]: Can this create issue if the user is operating over a hostname and the cors has ip addresses set. Can you check that usecase?

Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py]: Thanks @cloudbehl for pointing out, To cover this scenario we need to set the cross origin url with IP url and the fqdn as well. updated the PR accordingly.

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 2]
```diff
@@ -101,6 +105,26 @@ def auth(self, url: str, cluster_alias: str, username: str,
...
+        cors_endpoints_set = set()
+        cors_endpoints_set.add(hub_url)
+
+        orch = OrchClient.instance()
+        inventory_hosts = [host.to_json() for host in orch.hosts.list()]
+
+        for host in inventory_hosts:
+            host_addr = host['addr']
+            host_ip_url = hub_url.replace(hostname, host_addr)
+            host_hostname_url = hub_url.replace(hostname, host['hostname'])
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:120]: this will break ceph-dev like environments right? where no orchestrator is available?

Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py:120]: @nizamial09 , works fine with ceph-dev as well

**Review by nizamial09 — APPROVED** [review]

**Comment by cloudbehl** [comment]
jenkins test make check

**Pull request merged**